Tasks/FtpUploadV2/ftpuploadtask.ts

```diff
@@ -0,0 +1,354 @@
+import * as tl from "azure-pipelines-task-lib/task";
+import * as ftp from "basic-ftp";
+import * as fs from "fs";
+import * as path from "path";
+import * as url from "url";
+import { MatchOptions } from "azure-pipelines-task-lib/task";
+
+interface FtpOptions {
+    // url
+    serverEndpointUrl: url.Url;
+
+    // credentials
+    username: string;
+    password: string;
+
+    // other standard options
+    rootFolder: string;
+    filePatterns: string[];
+    remotePath: string;
+
+    // advanced options
+    clean: boolean;
+    cleanContents: boolean;
+    preservePaths: boolean;
+    trustSSL: boolean;
+}
+
+class ProgressTracker {
+    progressFilesUploaded: number = 0;
+    progressDirectoriesProcessed: number = 0;
+
+    constructor(private ftpOptions: FtpOptions, private fileCount: number) {
+        this.ftpOptions = ftpOptions;
+        this.fileCount = fileCount;
+    }
+
+    directoryProcessed(name: string): void {
+        this.progressDirectoriesProcessed++;
+        this.printProgress(
+            "remote directory successfully created/verified: " + name
+        );
+    }
+
+    fileUploaded(file: string, remoteFile: string): void {
+        this.progressFilesUploaded++;
+        this.printProgress("successfully uploaded: " + file + " to: " + remoteFile);
+    }
+
+    printProgress(message: string): void {
+        const total: number =
+            this.progressFilesUploaded + this.progressDirectoriesProcessed;
+        const remaining: number = this.fileCount - total;
+        console.log(
+            "files uploaded: " + this.progressFilesUploaded +
+            ", directories processed: " + this.progressDirectoriesProcessed +
+            ", total: " + total +
+            ", remaining: " + remaining +
+            ", " + message
+        );
+    }
+
+    getSuccessStatusMessage(): string {
+        return (
+            "\nhost: " + this.ftpOptions.serverEndpointUrl.host +
+            "\npath: " + this.ftpOptions.remotePath +
+            "\nfiles uploaded: " + this.progressFilesUploaded +
+            "\ndirectories processed: " + this.progressDirectoriesProcessed
+        );
+    }
+
+    getFailureStatusMessage() {
+        const total: number =
+            this.progressFilesUploaded + this.progressDirectoriesProcessed;
+            const remaining: number = this.fileCount - total;
+        return (
+            this.getSuccessStatusMessage() +
+            "\nunprocessed files & directories: " +
+            remaining
+        );
+    }
+}
+
+function getFtpOptions(): FtpOptions {
+    const options: FtpOptions = {} as FtpOptions;
```

Comment:
I think we should make the properties on `FtpOptions` optional. Then you shouldn't need this cast.